src/platform_impl/linux/wayland/seat/pointer/mod.rs

```diff
@@ -467,9 +469,107 @@ impl Dispatch<WpCursorShapeManagerV1, GlobalData, WinitState> for SeatState {
     }
 }
 
+pub struct PointerGesturesState {
+    pointer_gestures: ZwpPointerGesturesV1,
+}
+
+impl PointerGesturesState {
+    pub fn new(
+        globals: &GlobalList,
+        queue_handle: &QueueHandle<WinitState>,
+    ) -> Result<Self, BindError> {
+        let pointer_gestures = globals.bind(queue_handle, 1..=1, GlobalData)?;
+        Ok(Self { pointer_gestures })
+    }
+}
+
+impl Deref for PointerGesturesState {
+    type Target = ZwpPointerGesturesV1;
+
+    fn deref(&self) -> &Self::Target {
+        &self.pointer_gestures
+    }
+}
+
+impl Dispatch<ZwpPointerGesturesV1, GlobalData, WinitState> for PointerGesturesState {
+    fn event(
+        _state: &mut WinitState,
+        _proxy: &ZwpPointerGesturesV1,
+        _event: <ZwpPointerGesturesV1 as wayland_client::Proxy>::Event,
+        _data: &GlobalData,
+        _conn: &Connection,
+        _qhandle: &QueueHandle<WinitState>,
+    ) {
+        unreachable!("zwp_pointer_gestures_v1 has no events")
+    }
+}
+
+impl Dispatch<ZwpPointerGesturePinchV1, GlobalData, WinitState> for PointerGesturesState {
+    fn event(
+        state: &mut WinitState,
+        _proxy: &ZwpPointerGesturePinchV1,
+        event: <ZwpPointerGesturePinchV1 as wayland_client::Proxy>::Event,
+        _data: &GlobalData,
+        _conn: &Connection,
+        _qhandle: &QueueHandle<WinitState>,
+    ) {
+        let device_id = crate::event::DeviceId(crate::platform_impl::DeviceId::Wayland(DeviceId));
+        let (phase, pan_delta, scale_delta, rotation_delta) = match event {
+            Event::Begin { time: _, serial: _, surface, fingers } => {
+                if fingers != 2 {
+                    // We only support two fingers for now.
+                    return;
+                }
+                // The parent surface.
+                let parent_surface = match surface.data::<SurfaceData>() {
+                    Some(data) => data.parent_surface().unwrap_or(&surface),
+                    None => return,
+                };
+
+                let pan_delta = PhysicalPosition::new(0., 0.);
+                state.window_id = wayland::make_wid(parent_surface);
+                state.previous_scale = 1.;
+
+                (TouchPhase::Started, pan_delta, 0., 0.)
+            },
+            Event::Update { time: _, dx, dy, scale, rotation } => {
+                let pan_delta = PhysicalPosition::new(dx as f32, dy as f32);
```

Comment:
It's in `surface` coordinates, but `PhysicalPosition` is buffer, you should convert from logical to physical with the function like in `wl_pointer` events above.